src/Task/TalkNPC.pm

```diff
@@ -70,33 +80,107 @@ sub new {
 	my %args = @_;
 	my $self = $class->SUPER::new(@_, mutexes => MUTEXES);
 
+	$self->{type} = $args{type};
 	$self->{x} = $args{x};
 	$self->{y} = $args{y};
 	$self->{nameID} = $args{nameID};
 	$self->{sequence} = $args{sequence};
 	$self->{sequence} =~ s/^ +| +$//g;
-
-	# Watch for map change events. Pass a weak reference to ourselves in order
-	# to avoid circular references (memory leaks).
+	$self->{trying_to_cancel} = 0;
+	$self->{sent_talk_response_cancel} = 0;
+	$self->{wait_for_answer} = 0;
+	$self->{error_code} = undef;
+	$self->{error_message} = undef;
+	
 	my @holder = ($self);
 	Scalar::Util::weaken($holder[0]);
-	$self->{mapChangedHook} = Plugins::addHook('Network::Receive::map_changed', \&mapChanged, \@holder);
-
+	
+	push @{$self->{hookHandles}}, Plugins::addHooks(
+		['npc_talk',                  \&handle_npc_talk, \@holder],
+		['packet/npc_talk_continue',  \&handle_npc_talk, \@holder],
+		['npc_talk_done',             \&handle_npc_talk, \@holder],
+		['npc_talk_responses',        \&handle_npc_talk, \@holder],
+		['packet/npc_talk_number',    \&handle_npc_talk, \@holder],
+		['packet/npc_talk_text',      \&handle_npc_talk, \@holder],
+		['packet/npc_store_begin',    \&handle_npc_talk, \@holder],
+		['packet/npc_store_info',     \&handle_npc_talk, \@holder],
+		['packet/npc_sell_list',      \&handle_npc_talk, \@holder]
+	);
+	
 	return $self;
 }
 
+sub handle_npc_talk {
+	my ($hook_name, $args, $holder) = @_;
+	my $self = $holder->[0];
+	if ($hook_name eq 'npc_talk_done') {
+		$self->{stage} = AFTER_NPC_CLOSE;
+	}
+	$self->{time} = time;
+	$self->{sent_talk_response_cancel} = 0;
+	$self->{wait_for_answer} = 0;
+}
+
 sub DESTROY {
 	my ($self) = @_;
-	Plugins::delHook($self->{mapChangedHook}) if $self->{mapChangedHook};
+	debug "Destroying talknpc object\n", "ai_npcTalk";
+	Plugins::delHooks($_) for @{$self->{hookHandles}};
+	$self->SUPER::DESTROY;
 }
 
 # Overrided method.
 sub activate {
 	my ($self) = @_;
 	$self->SUPER::activate(); # Do not forget to call this!
 	$self->{time} = time;
-	$self->{stage} = 'Not Started';
-	$self->{mapChanged} = 0;
+	$self->{stage} = NOT_STARTED;
+}
+
+sub find_and_set_target {
+	my ($self) = @_;
+	my $target = $self->findTarget($npcsList);
+	if ($target) {
+		debug "Target NPC " . $target->name() . " at ($target->{pos}{x},$target->{pos}{y}) found.\n", "ai_npcTalk";
```

Comment:
If `$target` is an `Actor`, you can just do `"Target $target at..."` to get a suitable string representation (with actor type, binID and possibly other info). This can be used in translated strings too: `TF("Found %s", $target)`.